File: l-lang/src/bug.zig
pub const tokens_f = @import("tokens.zig");
const std = @import("std");
pub const Token = tokens_f.Token;
pub const TokenType = tokens_f.TokenType;
pub const Alloc = std.mem.Allocator;
const Arena = std.heap.ArenaAllocator;
const ArrayList = std.ArrayList;
const parser = @import("parser.zig");
const Expr = parser.Expr;
const utils = @import("utils.zig");
const Stream = utils.stream;
pub const Function = struct {
    args: std.ArrayList(Field),
    instructions: std.ArrayList(Statement),
    name: []const u8,
    rv: *Type,
    pub fn print(this: *const @This()) void {
        std.debug.print("{s}\nrv:", .{this.name});
        this.rv.print();
        std.debug.print("args:\n", .{});
        for (this.args.items) |i| {
            std.debug.print("{s}:", .{i.name});
            i.vtype.print();
        }
        std.debug.print("statements:\n", .{});
        for (this.instructions.items) |i| {
            i.print();
        }
        std.debug.print("end\n", .{});
    }
};
pub const Scope = struct { vars: std.ArrayList(Field), parent: ?*Scope };
pub const Parser = struct {
    functions: ArrayList(Function),
    types: ArrayList(Type),
    scope: ?*Scope,
    alloc: std.mem.Allocator,
    pub fn push_scope(this: *@This()) !void {
        var scope = try this.alloc.create(Scope);
        scope.parent = this.scope;
        scope.vars = try ArrayList(Field).initCapacity(this.alloc, 16);
        this.scope = scope;
    }
    pub fn pop_scope(this: *@This()) !void {
        if (this.scope == null) {
            return error.popped_scope_when_there_is_none;
        }
        const v = this.scope.?;
        this.scope = v.parent;
        this.alloc.destroy(v);
    }
};

pub const Field = struct {
    vtype: *Type,
    name: []const u8,
    offset: usize,
    size: usize,
};
pub const Type = union(enum) {
    t_int,
    t_flt,
    t_str,
    t_struct: struct {
        name: []const u8,
        fields: []Field,
    },
    pub fn print(this: *const @This()) void {
        std.debug.print("{s}\n", .{@tagName(this.*)});
    }
};
pub const Statement = union(enum) {
    basicExpr: Expr,
    if_statement: struct { cond: Expr, instructions: []Statement, elsef: []Statement },
    while_loop: struct { cond: Expr, instructions: []Statement },
    declare: struct {
        varname: []const u8,
        vtype: *Type,
        expr: Expr,
    },
    pub fn print(this: *const @This()) void {
        switch (this.*) {
            .basicExpr => |*expr| {
                std.debug.print("statement:", .{});
                expr.print();
            },
            .if_statement => |*f| {
                std.debug.print("if:", .{});
                f.cond.print();
                std.debug.print("do:", .{});
                for (f.instructions) |i| {
                    i.print();
                }
                std.debug.print("else:\n", .{});
                for (f.elsef) |i| {
                    i.print();
                }
                std.debug.print("end\n", .{});
            },
            .while_loop => |*p| {
                std.debug.print("while:\n", .{});
                p.cond.print();
                std.debug.print("do:\n", .{});
                for (p.instructions) |i| {
                    i.print();
                }
                std.debug.print("end\n", .{});
            },
            .declare => |*x| {
                std.debug.print("let {s}:", .{x.varname});
                x.vtype.print();
                std.debug.print("then:\n", .{});
                x.expr.print();
                std.debug.print("end\n", .{});
            },
        }
    }
};
pub const Program = struct {
    functions: ArrayList(Function),
    types: ArrayList(Type),
    pub fn print(this: *const @This()) void {
        for (this.types.items) |i| {
            i.print();
        }
        for (this.functions.items) |i| {
            i.print();
        }
    }
};
pub fn token_equals(l: Token, r: Token) bool {
    return l.tt == r.tt;
}
pub fn done_paren(str: *const Stream(Token)) bool {
    if (str.count_matches(TokenType.TokenOpenParen, TokenType.TokenCloseParen, token_equals)) {
        return true;
    }
    return false;
}
pub fn done_curly(str: *const Stream(Token)) bool {
    if (str.count_matches(TokenType.TokenOpenCUrly, TokenType.TokenCloseCurl, token_equals)) {
        return true;
    }
    return false;
}
pub fn done_semi(str: *const Stream(Token)) bool {
    str.contains(TokenType.TokenSemi, token_equals);
}
pub fn parse_statement(parse: Parser, s: Stream(Token)) !Statement {
    _ = s;
    _ = parse;
    return error.Unimplemented;
}
pub fn parse_scope(parse: Parser, s: Stream(Token)) !ArrayList(Statement) {
    _ = s;
    _ = parse;
    return error.Unimplemented;
}
pub fn parse_type(parse: Parser, s: Token) !*Type {
    _ = s;
    _ = parse;
    return Type.t_int;
}
pub fn parse_token_stream(alloc: Alloc, tokens: []Token) anyerror!Program {
    var stream = Stream(Token).new(tokens);
    var parse = Parser{ .alloc = alloc, .functions = try ArrayList(Function).initCapacity(alloc, 64), .scope = null, .types = try ArrayList(Type).initCapacity(alloc, 64) };
    while (true) {
        const sp = stream.try_next();
        if (sp == null) {
            break;
        }
        const s = sp.?;
        if (s.equals("defun")) {
            const name = try stream.next();
            const vt = try stream.next();
            if ((try stream.peek()).tt != TokenType.TokenOpenParen) {
                return error.expected_paren;
            }
            var expr = stream.collect(done_paren);
            var args = try ArrayList(Field).initCapacity(alloc, 16);
            _ = expr.try_next();
            while (true) {
                const arg_name = (try expr.next()).str;
                const arg_type = try parse_type(parse, try expr.next());
                if ((try expr.peek()).tt == TokenType.TokenCloseParen) {
                    break;
                }
                args.append(alloc, Field{ .name = arg_name, .vtype = arg_type, .offset = 0, .size = 8 });
            }
            const scope = stream.collect(done_curly);
            const scope_v = try parse_scope(scope);
            var f = Function{};
            f.args = args;
            f.instructions = scope_v;
            f.name = name;
            f.rv = parse_type(parse, vt);
            try parse.functions.append(alloc, f);
        } else if (s.equals("deftype")) {
            const name = try stream.next();
            _ = name;
            var scope = stream.collect(done_curly);
            _ = scope.next();
        }
    }
    return error.bruh;
}

File: bug/src/bug.rs
use crate::parser::Expr;
use crate::tokens::Token;
use crate::tokens::TokenStream;
use crate::tokens::TokenType;
use crate::utils::Stream;

use std::error::Error;
use std::sync::Arc;
#[derive(Clone, Debug, PartialEq)]
pub struct Var {
    pub vtype: Type,
    pub name: Arc<str>,
}
#[derive(Clone, Debug, PartialEq)]
pub enum Type {
    TStr,
    TInt,
    TChar,
    TFloat,
    TVoid,
    TStruct { name: Arc<str>, fields: Arc<[Var]> }, //ptr to struct,
    TFunction{return_type:Box<Type>, args:Vec<Type>},
    TPtr{tname:Arc<str>},
}

#[derive(Clone, Debug)]
pub enum Statement {
    While {
        cond: Expr,
        list: Vec<Statement>,
    },
    If {
        cond: Expr,
        list: Vec<Statement>,
        else_list: Vec<Statement>,
    },
    Declare {
        v: Var,
        list: Expr,
    },
    Basic {
        list: Expr,
    },
    Return {
        list: Expr,
    },
}
#[derive(Clone, Debug)]
pub struct Function {
    pub name: String,
    pub args: Vec<Var>,
    pub return_type: Type,
    pub list: Vec<Statement>,
}

#[derive(Clone, Debug)]
pub struct Context {
    pub types: Vec<Type>,
    pub functions: Vec<Function>,
    pub externs:Vec<Function>,
}
pub struct Scope {
    pub vars: Vec<Var>,
}
impl Context {}
pub fn parse_statement(
    context: &mut Context,
    tokens: &mut Stream<Token>,
) -> Result<Statement, Box<dyn Error>> {
    let base = tokens.peek().unwrap();
    if base.equals("if") {
        let _ = tokens.next();
        let exp = get_paren_expr(tokens);
        let expr = crate::parser::parse_expression(exp.values)?;
        println!("{:#?}", tokens);
        let scpe = get_scope(tokens);
        let scope = parse_scope(context, scpe)?;
        let mut elscp = Vec::new();
        if let Some(n) = tokens.peek() {
            if n.equals("else") {
                let _ = tokens.next();
                let nscpe = get_scope(tokens);
                elscp = parse_scope(context, nscpe)?;
            }
        }
        let out = Statement::If {
            cond: expr,
            list: scope,
            else_list: elscp,
        };
        return Ok(out);
    } else if base.equals("while") {
        let _ = tokens.next();
        let exp = get_paren_expr(tokens);
        let expr = crate::parser::parse_expression(exp.values)?;
        let scpe = get_scope(tokens);
        let scope = parse_scope(context, scpe)?;
        let out = Statement::While {
            cond: expr,
            list: scope,
        };
        return Ok(out);
    } else if base.equals("let") {
        _ = tokens.next().unwrap();
        let tt = tokens.next().unwrap();
        let vt = parse_type(&context, &tt.st)?;
        let namet = tokens.peek().unwrap();
        let name = namet.st;
        let mut ts = Vec::new();
        while let Some(k) = tokens.next() {
            if k.tt == TokenType::TokenSemi {
                break;
            }
            ts.push(k);
        }
        let expr = crate::parser::parse_expression(&ts)?;
        let out = Statement::Declare {
            v: Var { vtype: vt, name },
            list: expr,
        };
        return Ok(out);
    } else if base.equals("return") {
        _ = tokens.next().unwrap();

        let mut ts = Vec::new();
        while let Some(k) = tokens.next() {
            if k.tt == TokenType::TokenSemi {
                break;
            }
            ts.push(k);
        }
        let expr = crate::parser::parse_expression(&ts)?;
        return Ok(Statement::Return { list: expr });
    }
    let mut ts = Vec::new();
    while let Some(k) = tokens.next() {
        if k.tt == TokenType::TokenSemi {
            break;
        }
        ts.push(k);
    }
    let expr = crate::parser::parse_expression(&ts)?;
    let out = Statement::Basic { list: expr };
    Ok(out)
}
pub fn parse_scope(
    context: &mut Context,
    tokens: Stream<Token>,
) -> Result<Vec<Statement>, Box<dyn Error>> {
    let mut toks = tokens;
    let mut out = Vec::new();
    loop {
        if toks.peek().is_none() {
            break;
        }
        out.push(parse_statement(context, &mut toks)?);
    }
    Ok(out)
}
pub fn get_paren_expr<'a>(stream: &mut Stream<'a, Token>) -> Stream<'a, Token> {
    let mut count = 0;
    let tmps = stream.values;
    let mut i = 0;
    while let Some(s) = stream.next() {
        i += 1;
        if s.tt == TokenType::TokenOpenParen {
            count += 1;
        } else if s.tt == TokenType::TokenCloseParen {
            count -= 1;
        }
        if count == 0 {
            break;
        }
    }
    return Stream::new(&tmps[1..i - 1]);
}
pub fn parse_type(context: &Context, v: &str) -> Result<Type, Box<dyn Error>> {
    if v == "int" {
        return Ok(Type::TInt);
    } else if v == "float" {
        return Ok(Type::TFloat);
    } else if v == "string" {
        return Ok(Type::TStr);
    } else if v == "char" {
        return Ok(Type::TChar);
    }
    Err(format!("undefined type:{v}").into())
}
pub fn get_scope<'a>(stream: &mut Stream<'a, Token>) -> Stream<'a, Token> {
    let mut count = 0;
    let tmps = stream.values;
    let mut i = 0;
    while let Some(s) = stream.next() {
        i += 1;
        if s.tt == TokenType::TokenOpenCurl {
            count += 1;
        } else if s.tt == TokenType::TokenCloseCurl {
            count -= 1;
        }
        if count == 0 {
            break;
        }
    }
    return Stream::new(&tmps[1..i - 1]);
}
pub fn parse_program(tokens: Vec<Token>) -> Result<Context, Box<dyn Error>> {
    let mut context = Context {
        types: Vec::new(),
        functions: Vec::new(),
        externs:Vec::new()
    };
    let mut strem = Stream::new(&tokens);
    let old = strem.clone();
    while let Some(n) = strem.next() {
        if n.equals("deftype") {
            todo!();
        } else {
            continue;
        }
    }
    strem = old.clone();
    while let Some(n) = strem.next() {
        if n.equals("pub"){
            continue;
        }
        if n.equals("#include"){
            let to_import = strem.next().unwrap();
            let s= to_import.st.strip_prefix('"').unwrap().strip_suffix('"').unwrap();
            include_file(&mut context, s)?;

        }
        if n.equals("extern") {
            let name = strem.next().unwrap();
            let rvt = strem.next().unwrap();
            let rv = parse_type(&context, &rvt.st)?;
            let mut args: Vec<Var> = Vec::new();
            let mut s = strem.next().unwrap();
            if s.tt != TokenType::TokenOpenParen {
                todo!();
            }
            while let Some(p) = strem.next() {
                if p.tt == TokenType::TokenCloseParen {
                    break;
                }
                let name = p.st.clone();
                let t = strem.next().unwrap();
                let vt = parse_type(&context, &t.st)?;
                let v = Var {
                    name: name,
                    vtype: vt,
                };
                args.push(v);
            }
            s = strem.next().unwrap();
            if s.tt != TokenType::TokenSemi {
                todo!();
            }   
            context.externs.push(Function { name:name.st.to_string(), args, return_type: rv, list:Vec::new() });
        } else {
            continue;
        }
    }
    strem = old;
    while let Some(n) = strem.next() {
        if n.equals("defun") {
            let name = strem.next().unwrap();
            let rvt = strem.next().unwrap();
            let rv = parse_type(&context, &rvt.st)?;
            let mut args: Vec<Var> = Vec::new();
            let s = strem.next().unwrap();
            if s.tt != TokenType::TokenOpenParen {
                todo!();
            }
            while let Some(p) = strem.next() {
                if p.tt == TokenType::TokenCloseParen {
                    break;
                }
                let name = p.st.clone();
                let t = strem.next().unwrap();
                let vt = parse_type(&context, &t.st)?;
                let v = Var {
                    name: name,
                    vtype: vt,
                };
                args.push(v);
            }
            let s = get_scope(&mut strem);
            let scope = parse_scope(&mut context, s)?;
            let f = Function {
                name: name.st.to_string(),
                args,
                return_type: rv,
                list: scope,
            };
            context.functions.push(f);
        }
    }
    Ok(context)
}


pub fn include_file(context:&mut Context, s:&str)->Result<(), Box<dyn Error>>{
    let f = std::fs::read_to_string(s)?;
    let mut tokens = TokenStream::new(&f, s);
    while let Some(n) = tokens.next(){
        if n.equals("pub"){
            let mut p = tokens.next().unwrap();
            if p.equals("extern"){
                p = tokens.next().unwrap();
            }
            if !p.equals("defun"){
                todo!();
            }
            let name = tokens.next().unwrap();
            let rvt = tokens.next().unwrap();
            let rv = parse_type(&context, &rvt.st)?;
            let mut args: Vec<Var> = Vec::new();
            let s = tokens.next().unwrap();
            if s.tt != TokenType::TokenOpenParen {
                todo!();
            }
            while let Some(p) = tokens.next() {
                if p.tt == TokenType::TokenCloseParen {
                    break;
                }
                let name = p.st.clone();
                let t = tokens.next().unwrap();
                let vt = parse_type(&context, &t.st)?;
                let v = Var {
                    name: name,
                    vtype: vt,
                };
                args.push(v);
            }
            let f = Function{name:name.st.to_string(),args, return_type:rv,list:Vec::new()};
            context.externs.push(f);
        }
    }
    Ok(())

}
const BUITLINS:&[(&str, usize)] = &[
    ("int_to_string", 0x10000000), 
    ("float_to_string", 0x10000001), 
    ("print", 0x10000002), 
    ("println", 0x10000003), 
    ("strcat", 0x10000004), 
    ("or", 0x10000005), 
    ("and", 0x10000006), 
];
pub fn builtin(s:&str)->Option<usize>{
    for i in BUITLINS{
        if i.0 == s{
            return Some(i.1);
        }
    }
    None
}